onnx_models/conftest.py

```diff
@@ -0,0 +1,243 @@
+    """Metadata for a single input or output used with iree-run-module tooling.
+
+    Args:
+        name: The name of the parameter.
+        type: The type of the parameter as expected by the tools, e.g. "2x2xi32".
+        data_file: Path to either the input or expected output binary file for this parameter.
+    """
+
+    name: str
+    type: str
+    data_file: Path
+
+
+@dataclass(frozen=True)
+class OnnxModelMetadata:
+    """Metadata for an ONNX model.
+
+    Args:
+        inputs: One parameter metadata per input.
+        outputs: One parameter metadata per output.
+    """
+
+    inputs: list[IreeModelParameterMetadata]
+    outputs: list[IreeModelParameterMetadata]
+
+
+def convert_onnxruntime_node_arg_to_numpy_dimensions(
+    node_arg: NodeArg,
+) -> tuple[int]:
+    # Note: turning dynamic dimensions into just 1 here, since we need
+    # a concrete (static) shape buffer of input data in the tests.
+    return tuple(x if isinstance(x, int) else 1 for x in node_arg.shape)
+
+
+def convert_onnxruntime_shape_to_iree_type_string(
+    node_arg: NodeArg,
+) -> str:
+    # Note: turning dynamic dimensions into just "1" here, since we need
+    # a concrete (static) shape buffer of input data in the tests.
+    shape = "x".join([str(x) if isinstance(x, int) else "1" for x in node_arg.shape])
+    dtype = convert_node_arg_type_to_iree_dtype(node_arg.type)
+    if shape == "":
+        return dtype
+    return f"{shape}x{dtype}"
+
+
+def get_onnx_model_metadata(onnx_path: Path) -> OnnxModelMetadata:
+    # We can either
+    #   A) List all metadata explicitly
+    #   B) Get metadata on demand from the .onnx protobuf using 'onnx'
+    #   C) Get metadata on demand from the InferenceSession using 'onnxruntime'
+    # This is option C.
+
+    onnx_session = InferenceSession(onnx_path)
+    logger.info(f"Getting model metadata for '{onnx_path.relative_to(THIS_DIR)}'")
+    inputs = []
+    onnx_inputs = {}
+    for idx, input in enumerate(onnx_session.get_inputs()):
+        logger.debug(f"Session input [{idx}]")
+        logger.debug(f"  name: '{input.name}'")
+        numpy_dimensions = convert_onnxruntime_node_arg_to_numpy_dimensions(input)
+        iree_type = convert_onnxruntime_shape_to_iree_type_string(input)
+        logger.debug(f"  shape: {input.shape}")
+        logger.debug(f"  numpy shape: {numpy_dimensions}")
+        logger.debug(f"  type: '{input.type}'")
+        logger.debug(f"  iree parameter: {iree_type}")
+
+        # Create a numpy tensor with some random data for the input.
+        numpy_dtype = convert_node_arg_type_to_numpy_dtype(input.type)
+        if numpy_dtype == np.float32 or numpy_dtype == np.float64:
+            input_data = rng.random(numpy_dimensions, dtype=numpy_dtype)
+        elif numpy_dtype == np.int32 or numpy_dtype == np.int64:
+            input_data = rng.integers(numpy_dimensions, dtype=numpy_dtype)
+        else:
+            raise NotImplementedError(f"Unsupported numpy type: {numpy_dtype}")
```

Comment:
I'm sure you've already looked at this, but could consider borrowing some of <https://github.com/nod-ai/SHARK-TestSuite/blob/dbbe625d5b7beac0b4465f7b90441d9f4593bc57/alt_e2eshark/e2e_testing/onnx_utils.py#L31>

Reply:
Especially since you aren't using the `numpy_dtype` elsewhere.

Reply:
Ah! I hadn't looked there recently. Thanks for the reminder. Will pull in some of the cases from there.

Reply:
Refactored a bit, pulling more code from there. Thanks again!